crates/circuit/src/dag_circuit.rs

```diff
@@ -2919,43 +2919,80 @@ def _format(operand):
 
     /// Remove all of the ancestor operation nodes of node.
     fn remove_ancestors_of(&mut self, node: &DAGNode) -> PyResult<()> {
-        // anc = rx.ancestors(self._multi_graph, node)
-        // # TODO: probably better to do all at once using
-        // # multi_graph.remove_nodes_from; same for related functions ...
-        //
-        // for anc_node in anc:
-        //     if isinstance(anc_node, DAGOpNode):
-        //         self.remove_op_node(anc_node)
-        todo!()
+        let ancestors: Vec<_> = core_ancestors(&self.dag, node.node.unwrap())
+            .filter(|next| {
+                next != &node.node.unwrap()
+                    && match self.dag.node_weight(*next) {
+                        Some(NodeType::Operation(_)) => true,
+                        _ => false,
+                    }
+            })
+            .collect();
+        for a in ancestors {
+            self.dag.remove_node(a);
+        }
+        Ok(())
     }
 
     /// Remove all of the descendant operation nodes of node.
     fn remove_descendants_of(&mut self, node: &DAGNode) -> PyResult<()> {
-        // desc = rx.descendants(self._multi_graph, node)
-        // for desc_node in desc:
-        //     if isinstance(desc_node, DAGOpNode):
-        //         self.remove_op_node(desc_node)
-        todo!()
+        let descendants: Vec<_> = core_descendants(&self.dag, node.node.unwrap())
+            .filter(|next| {
+                next != &node.node.unwrap()
+                    && match self.dag.node_weight(*next) {
+                        Some(NodeType::Operation(_)) => true,
+                        _ => false,
+                    }
+            })
+            .collect();
+        for d in descendants {
+            self.dag.remove_node(d);
+        }
+        Ok(())
     }
 
     /// Remove all of the non-ancestors operation nodes of node.
     fn remove_nonancestors_of(&mut self, node: &DAGNode) -> PyResult<()> {
-        // anc = rx.ancestors(self._multi_graph, node)
-        // comp = list(set(self._multi_graph.nodes()) - set(anc))
-        // for n in comp:
-        //     if isinstance(n, DAGOpNode):
-        //         self.remove_op_node(n)
-        todo!()
+        let ancestors: Vec<_> = core_ancestors(&self.dag, node.node.unwrap())
```

Comment:
I would recommend using `HashSet` instead of `Vec` here so you can use `HashSet::contains` instead of `find`. With `Vec::find`, you will be scanning the list of `ancestors` for every node in the graph.

This otherwise looks good to me.